### ISSUE TITLE
`android_aar` rule builds a fat aar

### ISSUE THREAD

**Issue by thalescm** [facebook/buck#2020]
The `android_aar` rule is putting inside `claases.jar` all `.class` files from the `android_library` rule deps, and not only the jar generated by the rule (which contain only the `.class` generated by compiling the source code).

**Comment by thalescm** [comment]
Turns out it is also putting all `res` and from dependencies as well

**Comment by asp2insp** [comment]
I believe this is WAI. According to [the documentation](https://buckbuild.com/rule/android_aar.html#deps) `deps` are automatically traversed and bundled into the AAR. If you wish to compile against items without including them in the final AAR, you can use `provided_deps` to provide a compile-time dependency.

**Comment by thalescm** [comment]
@asp2insp I disagree with your point. I'll try to explain the best way I can to see if you agree with me. According to the same documentation:
```
List of build targets whose corresponding compiled Java code, Android resources, and native libraries will be included in the AAR.
```
So transitive dependencies from an `android_library`, which were **not**  compiled in that target shouldn't be included.

Suppose I have the following BUCK file:
```
android_resource(
    name="res",
    deps = [
        'path_to_prebuilt_aar_1',
        'path_to_prebuilt_aar_2',
    ]
)

android_library(
    name="lib",
    srcs = [
         'MyClass1.java',
         'MyClass2.java',
    ]
    <other_configs>,
    deps = [
        'path_to_prebuilt_aar_1',
        'path_to_prebuilt_aar_2',
        ':res',
    ]
)

android_aar(
    name="aar",
    deps= [
        ':lib',
        ':res',
    ]
)
```
both `aar_1` and `aar_2` are transitive dependencies from `lib` and `res`, thus were not compiled in their respective build targets.
The resulting `aar.aar` will have all `.class` and all resources included from both `aar_1` and `aar_2`, which is wrong and don't respect the documentation.
The issue should be reopened.

**Comment by asp2insp** [comment]
I definitely agree we should clarify the documentation. Can you provide some more context about your use case? I believe we mostly use `android_aar` for generating SDKs and we definitely would like those to be "fat"

**Comment by thalescm** [comment]
Why would you like those to be "fat"? Transitive dependencies should caught and added only when building the APK.
When building this "fat" aar, not only you're increasing the size of your SDK but you're also possibly messing up with the SDK users build, as they might have the same dependencies, such as, for example, `android.support` libs.

I'm having strange runtime errors, such as methods not found when unit testing, and cannot build my apk due to resources collisions, such as below:
```
buck-out/bin/.okbuck/cache/__unpack_com.android.support.appcompat-v7-27.1.1.aar#aar_unzip__/res/values/values.xml:164: Original attribute defined here.
buck-out/bin/.okbuck/cache/__unpack_com.nubank.android.foundation.aar#aar_unzip__/res/values/values.xml:345: error: Attribute "showAsAction" already defined with incompatible format.
```
If I remove all transitive dependencies info (everything that is not originally from the direct deps of the aar rule) from the generated aar, everything works as expected.

**Comment by thalescm** [comment]
I'd say that we could add a configuration to `android_aar` that would tell buck if you wish or not to include those. WDYT?

**Comment by asp2insp** [comment]
Sorry, I didn't see your last comment! We probably won't take that on because the current behavior matches our needs, but I'm happy to review a PR if you put one up.

**Comment by kageiit** [comment]
@asp2insp this does not seem correct. `prebuilt_jar` does not have this behavior and it supports correctly packaging transitive deps and creating poms when running `buck publish`. `android_aar` should do the same (so its not WAI). I believe there is some work to make `android_aar` behave the same as `prebuilt_jar` for publishing. Ideally for fb's needs, there should be a separate option/rule to make the aar fat

**Comment by asp2insp** [comment]
@kageiit `prebuilt_jar` does not seem like an apt comparison, since it does not actually build anything. As written, an AAR is most similar to an `android_binary` rule. As I stated above, it would be a breaking change to make the aar skinny by default, but I will happily review a PR to add a flag to opt into this behavior

**Comment by kageiit** [comment]
I think only FB builds fat aars by default :)

But a flag would be the best option and we can force it to false for anyone using okbuck so that becomes the default. cc @thalescm

**Comment by thalescm** [comment]
Sorry guys, I was at holiday Friday so I didn't even opened my note.
I'll see if I can work on the flag to turn it skinny, and then create this DEF on okbuck as well
